File: Vysakh_HW 4/particle_swarm.m
% Vysakh Venugopal (M10665532) Assignment #4 - Problem 3

%%%%%%%%%%%%%%%%% PARTICLE SWARM OPTIMIZATION %%%%%%%%%%%%%%%%%%%%%
function xmin = particle_swarm(F,dimension,particles,maxloop,inertia,phip,phig,lb,ub)
%PARTICLE_SWARM finds global optimum of an objective function using particle swarm
%optimization method
% F: Objective Function
% dimension: Dimension of the problem
% particles: number of particles
% maxloop: Maximum number of iterations
% inertia: inertia coefficient
% phip: weighting coefficient for personal best position
% phig: weighting coefficient for global best position
% lb: lower bound of search space in each dimension
% ub: upper bound of search space in each dimension

searchSpace = horzcat(lb,ub);
x0 = zeros(particles,dimension);    % input initialization
v0 = zeros(particles,dimension);    % velocity initialization
for i = 1:dimension
    x0(:,i) = (searchSpace(i+dimension) - searchSpace(i)).*rand(particles,1) + searchSpace(i);
    v0(:,i) = rand(particles,1);
end
xp = x0;        % local best is set as initial input
Fp = zeros(particles,1);
for i = 1:particles
    Fp(i) = F(xp(i,:));
end
[~,idx] = min(Fp);      % finding minimum input
xg = xp(idx,:);
loop = 1;
while loop <= maxloop
    for i = 1:particles
        v0(i,:) = inertia*v0(i,:) + phip*rand(1)*(xp(i,:) - x0(i,:)) + phig*rand(1)*(xg - x0(i,:));     % velocity update
        x0(i,:) = x0(i,:) + v0(i,:);
        if F(x0(i,:)) < F(xp(i,:))
            xp(i,:) = x0(i,:);
            Fp(i) = F(xp(i,:));
        end
    end
    [~,idx] = min(Fp);
    xg = xp(idx,:);     % global best
    loop = loop + 1;
end
xmin = xg;
end

File: VJ_Assignment4/ParticleSwarm.py
# Created by Vinay Jakkali on 12/12/2019

"""
--------------------------------------------------------------------------------------
The particle_swarm function defined here takes following parameters as input

INPUT:
    F : It is the objective function

    Dims : It is the total # of dimensions

    Particles : It is the total # of particles

    Omega : Particle Swarms inertial motion

    Phi_p : Movement towards a particle's best known position

    Phi_g : Movement towards a swarm's best known position

    L : Lower bound of the search space

    U : Upper bound of the search space

    maxiter : Maximum iterations to stop the optimization
    (by default maxiter = 1e3)

OUTPUT:
    1. Returns the values of minimum 'Xk_min'
--------------------------------------------------------------------------------------
"""
import numpy as np
from Rosenbrock import rosenbrock

def particle_swarm(F, Dims, Particles, Omega, Phi_p, Phi_g, L, U, maxiter=1e3):

    """ Particle Swarm Optimization for a given objective function """

    """ Creating the search space """

    Low = np.full(Dims, L)

    Up = np.full(Dims, U)

    SearchGrid = np.hstack((Low, Up))

    """ Initializing position and velocity matrices """

    X0 = np.zeros((Particles, Dims))
    V0 = np.zeros((Particles, Dims))

    for i in range(Dims):

        """ Randomly initializing particles' position """

        X0[:, i] = (SearchGrid[i+Dims] - SearchGrid[i]) * np.random.rand(Particles) + SearchGrid[i]

        V0[:, i] = np.random.rand(Particles)  # Randomly initializing each particle with a velocity

    X_local = X0  # Initializing the best known position for each particle

    F_IP = np.zeros(Particles)  # Initializing objective at each particle's position

    for i in range(Particles):

        F_IP[i] = F(X_local[i, :])  # Computing objective at each particle's position

    index = np.argmin(F_IP)  # Finding the particle with smallest objective value

    X_global = X_local[index, :]  # Setting Swarm's best known position

    k = 0

    while k < maxiter:

        for i in range(Particles):

            """ Updating particles' velocity """

            V0[i, :] = Omega * V0[i, :] + Phi_p * np.random.rand(1) * (X_local[i, :] - X0[i, :]) + \
                Phi_g * np.random.rand(1) * (X_global - X0[i, :])

            X0[i, :] = X0[i, :] + V0[i, :]  # Updating particles' position

            if F(X0[i, :]) < F(X_local[i, :]):  # Updating each particles' best known position

                X_local[i, :] = X0[i, :]

                F_IP[i] = F(X_local[i, :])

        index = np.argmin(F_IP)

        X_global = X_local[index, :]  # Updating Swarm's best known position

        k += 1

    Xk_min = X_global

    return Xk_min


if __name__ == "__main__":


    xmin = particle_swarm(rosenbrock, 5, 500, 0.5, 2, 2, 0, 2)

    print("Xmin", xmin)

